File: java/client/src/main/java/glide/api/models/commands/WeightAggregateOptions.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.commands;

import static glide.utils.ArrayTransformUtils.concatenateArrays;

import glide.api.commands.SortedSetBaseCommands;
import glide.api.models.GlideString;
import java.util.ArrayList;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.apache.commons.lang3.tuple.Pair;

/**
 * Arguments for {@link SortedSetBaseCommands#zunion}, {@link
 * SortedSetBaseCommands#zunionWithScores}, {@link SortedSetBaseCommands#zinter}, {@link
 * SortedSetBaseCommands#zinterWithScores}, {@link SortedSetBaseCommands#zunionstore}, and {@link
 * SortedSetBaseCommands#zinterstore}.
 *
 * @see <a href="https://valkey.io/commands/zunion/">valkey.io</a> for more details.
 * @see <a href="https://valkey.io/commands/zinter/">valkey.io</a> for more details.
 * @see <a href="https://valkey.io/commands/zinterstore/">valkey.io</a> for more details.
 * @see <a href="https://valkey.io/commands/zunionstore/">valkey.io</a> for more details.
 */
public abstract class WeightAggregateOptions {
    public static final String WEIGHTS_VALKEY_API = "WEIGHTS";
    public static final String AGGREGATE_VALKEY_API = "AGGREGATE";

    /**
     * Option for the method of aggregating scores from multiple sets. This option defaults to SUM if
     * not specified.
     */
    public enum Aggregate {
        /** Aggregates by summing the scores of each element across sets. */
        SUM,
        /** Aggregates by selecting the minimum score for each element across sets. */
        MIN,
        /** Aggregates by selecting the maximum score for each element across sets. */
        MAX;

        public String[] toArgs() {
            return new String[] {AGGREGATE_VALKEY_API, toString()};
        }
    }

    /**
     * Basic interface. Please use one of the following implementations:
     *
     * <ul>
     *   <li>{@link KeyArray}
     *   <li>{@link WeightedKeys}
     * </ul>
     */
    public interface KeysOrWeightedKeys {
        /** Convert to command arguments according to the Redis API. */
        String[] toArgs();
    }

    /** Represents the keys of the sorted sets involved in the aggregation operation. */
    @RequiredArgsConstructor
    public static class KeyArray implements KeysOrWeightedKeys {
        private final String[] keys;

        @Override
        public String[] toArgs() {
            return concatenateArrays(new String[] {Integer.toString(keys.length)}, keys);
        }
    }

    /**
     * Represents the mapping of sorted set keys to their score weights. Each weight is used to boost
     * the scores of elements in the corresponding sorted set by multiplying them before their scores
     * are aggregated.
     */
    @RequiredArgsConstructor
    public static class WeightedKeys implements KeysOrWeightedKeys {
        private final List<Pair<String, Double>> keysWeights;

        @Override
        public String[] toArgs() {
            List<String> keys = new ArrayList<>();
            List<Double> weights = new ArrayList<>();
            List<String> argumentsList = new ArrayList<>();

            for (Pair<String, Double> entry : keysWeights) {
                keys.add(entry.getLeft());
                weights.add(entry.getRight());
            }
            argumentsList.add(Integer.toString(keys.size()));
            argumentsList.addAll(keys);
            argumentsList.add(WEIGHTS_VALKEY_API);
            for (Double weight : weights) {
                argumentsList.add(weight.toString());
            }

            return argumentsList.toArray(new String[0]);
        }
    }

    /**
     * Basic interface. Please use one of the following implementations:
     *
     * <ul>
     *   <li>{@link KeyArray}
     *   <li>{@link WeightedKeys}
     * </ul>
     */
    public interface KeysOrWeightedKeysBinary {
        /** Convert to command arguments according to the Redis API. */
        GlideString[] toArgs();
    }

    /** Represents the keys of the sorted sets involved in the aggregation operation. */
    @RequiredArgsConstructor
    public static class KeyArrayBinary implements KeysOrWeightedKeysBinary {
        private final GlideString[] keys;

        @Override
        public GlideString[] toArgs() {
            return concatenateArrays(new GlideString[] {GlideString.of(keys.length)}, keys);
        }
    }

    /**
     * Represents the mapping of sorted set keys to their score weights. Each weight is used to boost
     * the scores of elements in the corresponding sorted set by multiplying them before their scores
     * are aggregated.
     */
    @RequiredArgsConstructor
    public static class WeightedKeysBinary implements KeysOrWeightedKeysBinary {
        private final List<Pair<GlideString, Double>> keysWeights;

        @Override
        public GlideString[] toArgs() {
            List<GlideString> keys = new ArrayList<>();
            List<Double> weights = new ArrayList<>();
            List<GlideString> argumentsList = new ArrayList<>();

            for (Pair<GlideString, Double> entry : keysWeights) {
                keys.add(entry.getLeft());
                weights.add(entry.getRight());
            }
            argumentsList.add(GlideString.of(keys.size()));
            argumentsList.addAll(keys);
            argumentsList.add(GlideString.of(WEIGHTS_VALKEY_API));
            for (Double weight : weights) {
                argumentsList.add(GlideString.of(weight));
            }

            return argumentsList.toArray(new GlideString[0]);
        }
    }
}

File: go/api/options/weight_aggregate_options.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

package options

import "github.com/valkey-io/valkey-glide/go/utils"

// Aggregate represents the method of aggregating scores from multiple sets
type Aggregate string

const (
	AggregateSum Aggregate = "SUM" // Aggregates by summing the scores of each element across sets
	AggregateMin Aggregate = "MIN" // Aggregates by taking the minimum score of each element across sets
	AggregateMax Aggregate = "MAX" // Aggregates by taking the maximum score of each element across sets
)

// converts the Aggregate to its Valkey API representation
func (a Aggregate) ToArgs() []string {
	return []string{AggregateKeyWord, string(a)}
}

// This is a basic interface. Please use one of the following implementations:
// - KeyArray
// - WeightedKeys
type KeysOrWeightedKeys interface {
	ToArgs() []string
}

// represents a list of keys of the sorted sets involved in the aggregation operation
type KeyArray struct {
	Keys []string
}

// converts the KeyArray to its Valkey API representation
func (k KeyArray) ToArgs() []string {
	args := []string{utils.IntToString(int64(len(k.Keys)))}
	args = append(args, k.Keys...)
	return args
}

type KeyWeightPair struct {
	Key    string
	Weight float64
}

// represents the mapping of sorted set keys to their score weights
type WeightedKeys struct {
	KeyWeightPairs []KeyWeightPair
}

// converts the WeightedKeys to its Valkey API representation
func (w WeightedKeys) ToArgs() []string {
	keys := make([]string, 0, len(w.KeyWeightPairs))
	weights := make([]string, 0, len(w.KeyWeightPairs))
	args := make([]string, 0)
	for _, pair := range w.KeyWeightPairs {
		keys = append(keys, pair.Key)
		weights = append(weights, utils.FloatToString(pair.Weight))
	}
	args = append(args, utils.IntToString(int64(len(keys))))
	args = append(args, keys...)
	args = append(args, WeightsKeyword)
	args = append(args, weights...)
	return args
}
